File: tools/wavcmp.mjs
// How different are two WAVs, really?
//
// The emulator does not reproduce a real CPU bit for bit and is not trying to:
// RSQRTPS and RCPPS are *approximate* by definition, hardware answers them to
// about twelve bits, and this emulator computes them exactly - which is inside
// what the architecture allows and outside what a byte comparison forgives.
// Native ONNX Runtime also picks AVX2 kernels where the emulated one is told it
// has only SSE2, and the two sum in different orders.
//
// So "identical" is the wrong question and this asks the right one: how far
// apart are the samples, against the peak of the signal.
//
//   node tools/wavcmp.mjs a.wav b.wav
//
// A difference of a few least-significant bits on a 16-bit signal is the same
// speech; anything you could hear would be percent, not thousandths.
import fs from 'node:fs';

const [, , pathA, pathB] = process.argv;
if (!pathA || !pathB) {
    console.error('usage: node tools/wavcmp.mjs a.wav b.wav');
    process.exit(2);
}

const a = fs.readFileSync(pathA);
const b = fs.readFileSync(pathB);
const HEADER = 44;  // canonical RIFF/WAVE, which is what CORE writes

if (a.length !== b.length)
    console.log(`lengths differ: ${a.length} vs ${b.length} bytes`);

const n = (Math.min(a.length, b.length) - HEADER) >> 1;
let maxDiff = 0, sumSq = 0, peak = 0, at = -1;
for (let i = 0; i < n; i++) {
    const x = a.readInt16LE(HEADER + i * 2);
    const y = b.readInt16LE(HEADER + i * 2);
    const d = Math.abs(x - y);
    if (d > maxDiff) { maxDiff = d; at = i; }
    sumSq += d * d;
    peak = Math.max(peak, Math.abs(x), Math.abs(y));
}
const rms = Math.sqrt(sumSq / n);

console.log(`${n} samples, peak ${peak}`);
console.log(`max difference ${maxDiff} (sample ${at}) = ${(100 * maxDiff / peak).toFixed(3)}% of peak`);
console.log(`rms difference ${rms.toFixed(3)} = ${(100 * rms / peak).toFixed(4)}% of peak`);

File: tools/wavcmp.py
"""Compare two 16-bit PCM WAV files sample for sample.

The same job as tools/wavcmp.mjs, for machines without node.  Prints the
largest difference against the peak, which is the number this project has been
quoting all along: "the same utterance, through four very different machines".

    python tools/wavcmp.py a.wav b.wav
"""
import struct
import sys


def read_wav(path):
    with open(path, "rb") as f:
        data = f.read()
    if data[:4] != b"RIFF" or data[8:12] != b"WAVE":
        raise SystemExit("%s: not a RIFF/WAVE file" % path)
    pos, fmt, pcm = 12, None, None
    while pos + 8 <= len(data):
        cid = data[pos:pos + 4]
        size = struct.unpack_from("<I", data, pos + 4)[0]
        body = data[pos + 8:pos + 8 + size]
        if cid == b"fmt ":
            fmt = struct.unpack_from("<HHIIHH", body, 0)
        elif cid == b"data":
            pcm = body
        pos += 8 + size + (size & 1)
    if fmt is None or pcm is None:
        raise SystemExit("%s: missing fmt or data chunk" % path)
    if fmt[5] != 16:
        raise SystemExit("%s: %d-bit, expected 16" % (path, fmt[5]))
    n = len(pcm) // 2
    return fmt, struct.unpack_from("<%dh" % n, pcm, 0)


def main():
    if len(sys.argv) != 3:
        raise SystemExit(__doc__)
    fa, a = read_wav(sys.argv[1])
    fb, b = read_wav(sys.argv[2])
    print("  %-28s %d ch, %d Hz, %d samples" % (sys.argv[1], fa[1], fa[2], len(a)))
    print("  %-28s %d ch, %d Hz, %d samples" % (sys.argv[2], fb[1], fb[2], len(b)))
    if fa[1] != fb[1] or fa[2] != fb[2]:
        print("  channel count or rate differs")
    n = min(len(a), len(b))
    if len(a) != len(b):
        print("  lengths differ by %d samples; comparing the first %d"
              % (abs(len(a) - len(b)), n))
    worst, at, differing = 0, -1, 0
    for i in range(n):
        d = abs(a[i] - b[i])
        if d:
            differing += 1
        if d > worst:
            worst, at = d, i
    peak = max(max((abs(v) for v in a[:n]), default=0),
               max((abs(v) for v in b[:n]), default=0))
    print("")
    print("  samples differing   %d of %d" % (differing, n))
    print("  largest difference  %d at sample %d" % (worst, at))
    print("  peak amplitude      %d" % peak)
    if peak:
        print("  worst as a fraction %.4f %%" % (100.0 * worst / peak))
    sys.exit(0 if worst == 0 else 1)


main()
